src/Processors/QueryPlan/Optimizations/optimizeReadInOrder.cpp

```diff
@@ -0,0 +1,934 @@
+
+    //std::cerr << "---- optimizeReadInOrder found sorting" << std::endl;
+
+    if (sorting->getType() != SortingStep::Type::Full)
+        return;
+
+    if (typeid_cast<UnionStep *>(node.children.front()->step.get()))
+    {
+        auto & union_node = node.children.front();
+
+        std::vector<InputOrderInfoPtr> infos;
+        const SortDescription * max_sort_descr = nullptr;
+        infos.reserve(node.children.size());
+        for (auto * child : union_node->children)
+        {
+            infos.push_back(buildInputOrderInfo(*sorting, *child));
+
+            if (infos.back() && (!max_sort_descr || max_sort_descr->size() < infos.back()->sort_description_for_merging.size()))
+                max_sort_descr = &infos.back()->sort_description_for_merging;
+        }
+
+        if (!max_sort_descr || max_sort_descr->empty())
+            return;
+
+        for (size_t i = 0; i < infos.size(); ++i)
+        {
+            const auto & info = infos[i];
+            auto & child = union_node->children[i];
+
+            QueryPlanStepPtr additional_sorting;
+
+            if (!info)
+            {
+                auto limit = sorting->getLimit();
+                /// If we have limit, it's better to sort up to full description and apply limit.
+                /// We cannot sort up to partial read-in-order description with limit cause result set can be wrong.
+                const auto & descr = limit ? sorting->getSortDescription() : *max_sort_descr;
+                additional_sorting = std::make_unique<SortingStep>(
+                    child->step->getOutputStream(),
+                    descr,
+                    limit, /// TODO: support limit with ties
+                    sorting->getSettings(),
+                    false);
+            }
+            else if (info->sort_description_for_merging.size() < max_sort_descr->size())
+            {
+                additional_sorting = std::make_unique<SortingStep>(
+                    child->step->getOutputStream(),
+                    info->sort_description_for_merging,
+                    *max_sort_descr,
+                    sorting->getSettings().max_block_size,
+                    0); /// TODO: support limit with ties
+            }
+
+            if (additional_sorting)
+            {
+                auto & sort_node = nodes.emplace_back();
+                sort_node.step = std::move(additional_sorting);
+                sort_node.children.push_back(child);
+                child = &sort_node;
+            }
+        }
+
+        sorting->convertToFinishSorting(*max_sort_descr);
+    }
+    else if (auto order_info = buildInputOrderInfo(*sorting, *node.children.front()))
+    {
+        sorting->convertToFinishSorting(order_info->sort_description_for_merging);
+    }
+}
+
+/// This optimisation is obsolete and will be removed.
+/// optimizeReadInOrder covers it.
+size_t tryReuseStorageOrderingForWindowFunctions(QueryPlan::Node * parent_node, QueryPlan::Nodes & /*nodes*/)
```

Comment:
Need to remove this. It should be supported by default from code above, and even better.

Reply:
I like to have it till next stable release just in case.